### PR TITLE
test(devnet): define RFC-64 Gate 2 exact-set evidence

### PR DESCRIPTION
**Pull request by branarakic**
## Scope\n\nAdds the closed deterministic evidence contract for RFC-64 Gate 2 multi-asset completeness.\n\nThis PR intentionally contains only the verifier/generator/schema package. It does not claim Gate 2 has passed and it does not modify product runtime code.\n\n## What it proves\n\n- exact 1..1024 asset set equality\n- exact authored head, receiver head, signed row count, and activated row count\n- exact per-asset KA ID, UAL, row/content/seal/bundle digests, and triple count\n- product-compatible catalog-scope and applied-inventory digest recomputation\n- deterministic RFC 8785 JCS plus one LF output\n- rejection of missing, extra, duplicate, malformed, accessor, Proxy, 0-row, and 1025-row evidence\n- fixture output remains explicitly not-connected and not-evaluated\n\n## Verification\n\n- pnpm test: 19 passed\n- pnpm typecheck: passed\n- git diff --check: passed\n\n## Integration\n\nBase: integration/rfc64-devnet\nNo RFC-64 work is targeted at main.

### CODE REVIEW

File: devnet/rfc64-gate2-multi-asset-completeness/src/product-digests.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+/**
+ * Exact Gate-1-compatible production applied-inventory framing. Rows are in
+ * numeric KA-ID order. Bundle digests are deliberately not inside this legacy
+ * commitment, so the contract separately requires exact full-row equality.
+ */
+export function computeAppliedInventoryDigest(
+  catalogScopeDigest: string,
+  inputRows: readonly AssetRowV1[],
+): string {
+  const rows = [...inputRows].sort(compareRowsByKaId);
```
Comment (otReviewAgent) [inline devnet/rfc64-gate2-multi-asset-completeness/src/product-digests.ts:29]: 🔴 Bug: Applied-inventory digest uses a different row order than the production receiver

**What's wrong**
The new contract says it is recomputing the production applied-inventory digest, but it hashes rows in mathematical `kaId` order. The current production receiver hashes rows in lexical `kaUal` order. For multi-digit KA numbers, those orders differ, so the fixture verifier can reject valid product evidence or produce a digest that no receiver will declare.

**Example**
A real receiver digest for rows with UALs ending `/2`, `/10`, and `/100` is computed in lexical UAL order by the production receiver. This verifier recomputes in numeric `kaId` order, so `inventoryDigestMatches` becomes false even when the product receiver produced correct applied-inventory evidence.

**Suggested direction**
Use the same row ordering as the production digest implementation, or update both sides together if numeric KA-ID order is the intended new contract.

**For Agents**
In `devnet/rfc64-gate2-multi-asset-completeness/src/product-digests.ts`, align `computeAppliedInventoryDigest` with the production `computeRfc64AppliedInventoryDigestV1` ordering or change the production contract intentionally in the same PR. Preserve the digest domain and row fields, and prove with a vector containing at least KA numbers 2, 10, and 100 that the helper matches the production receiver digest.

File: devnet/rfc64-gate2-multi-asset-completeness/src/product-digests.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  encodeU64,
+  sha256Digest,
+} from './canonical.ts';
+import type { AssetRowV1, CatalogScopeV1 } from './schema.ts';
+
+const CATALOG_SCOPE_DIGEST_DOMAIN = 'dkg-author-catalog-scope-v1\n';
+const APPLIED_INVENTORY_DIGEST_DOMAIN = 'dkg-rfc64-applied-inventory-v1\n';
+
+/** Exact dkg-core `computeAuthorCatalogScopeDigestV1` framing. */
+export function computeCatalogScopeDigest(scope: CatalogScopeV1): string {
```
Comment (otReviewAgent) [inline devnet/rfc64-gate2-multi-asset-completeness/src/product-digests.ts:16]: 🟡 Issue: Do not shadow the product serialization and digest layer inside the harness

**What's wrong**
The PR creates a parallel implementation of consensus-adjacent primitives: RFC 8785 canonical JSON, author-catalog scope digest framing, applied-inventory digest framing, digest parsing, and u64 encoding. That is a structural regression because the harness claims production compatibility while decoupling itself from the product code that owns those contracts. Any protocol change now has to be patched in two places, and drift will look like ordinary fixture code rather than an obvious compile-time dependency.

**Example**
This file sorts applied-inventory rows with compareRowsByKaId, while the existing production helper sorts rows by kaUal. Even if that difference is intentional for this fixture, keeping a second hand-written digest implementation forces every future reader to audit which one is canonical.

**Suggested direction**
Add the appropriate workspace dependency or expose the missing production helper from the owning package, then reduce this file to adapter-shaped projection code. If the harness needs stricter document framing, wrap the canonical helper rather than reimplementing RFC 8785, digest byte parsing, u64 encoding, and domain-separated hashing locally.

**For Agents**
Look at devnet/rfc64-gate2-multi-asset-completeness/src/product-digests.ts and src/canonical.ts. Preserve the emitted fixture/verdict bytes, but either depend on/export the existing core and receiver digest helpers or move the shared framing into one canonical module that both product and harness call. Keep a golden vector proving the harness still matches the intended product digest.

File: devnet/rfc64-gate2-multi-asset-completeness/src/schema.ts [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+  readonly extraKaIds: readonly string[];
+  readonly duplicateKaIds: readonly string[];
+  readonly duplicateUals: readonly string[];
+  readonly mismatchedKaIds: readonly string[];
+  readonly recomputedCatalogScopeDigest: string;
+  readonly recomputedInventoryDigest: string;
+  readonly rejectReasons: readonly string[];
+};
+
+const RAW_KEYS = ['authored', 'gateEvaluation', 'productBoundary', 'received', 'schema'];
```
Comment (otReviewAgent) [inline devnet/rfc64-gate2-multi-asset-completeness/src/schema.ts:105]: 🟡 Issue: Collapse the parallel contract declarations into one typed boundary model

**What's wrong**
The schema boundary is maintained through several disconnected representations: exported TypeScript types, expected-key arrays, manual validators, manual object reconstruction, and unchecked casts. That makes the contract harder to evolve safely and makes the most important invariant, exact artifact shape, depend on humans keeping parallel lists synchronized.

**Example**
Adding a new row field would require touching AssetRowV1, ROW_KEYS, parseRows validation, the row reconstruction, sameRow, and likely digest/verdict code. TypeScript does not prove ROW_KEYS is exhaustive or that each reconstructed field has the declared type; the casts at the boundary paper over that gap.

**Suggested direction**
Make the schema description the single source of truth for exact keys, validators, and typed construction. A tiny local parser combinator would delete the string-list bookkeeping and most of the casts while keeping the same strict boundary semantics.

**For Agents**
Start in src/schema.ts. Preserve the fail-closed exact-field behavior, accessor rejection, and row-count cap. Replace the parallel key arrays plus casts with a small typed schema/field descriptor helper, for example record({ kaId: canonicalU256, catalogRowDigest: digest32, ... }) that derives exact keys, snapshots descriptors, validates fields, and returns a typed frozen object. Add a focused regression that deleting or adding a field still fails closed.

File: devnet/rfc64-gate2-multi-asset-completeness/src/verify.ts [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+  VERDICT_SCHEMA_ID,
+  parseRawEvidence,
+  rowBindsToScope,
+  type AssetRowV1,
+  type CompletenessChecksV1,
+  type RawEvidenceV1,
+  type VerdictV1,
+} from './schema.ts';
+
+const CHECK_NAMES: readonly (keyof CompletenessChecksV1)[] = Object.freeze([
```
Comment (otReviewAgent) [inline devnet/rfc64-gate2-multi-asset-completeness/src/verify.ts:19]: 🟡 Issue: Make the verdict check registry exhaustive instead of hand-synchronized

**What's wrong**
The verifier introduces another set of parallel contract metadata. The check type, ordered names, reason map, actual checks object, fixtureComplete calculation, and schema-reject object all have to stay aligned manually. This is exactly the kind of small contract registry that tends to rot as soon as one more invariant is added.

**Example**
If a future check is added to CompletenessChecksV1 and the checks object but omitted from CHECK_NAMES, fixtureComplete and rejectReasons will silently ignore it because CHECK_NAMES is typed as readonly (keyof CompletenessChecksV1)[] rather than an exhaustive registry.

**Suggested direction**
Use one check definition table as the source for ordering, reason text, and completion/rejection derivation, or derive the names from an exact checks object with a typed helper. Avoid the Object.fromEntries cast in schemaRejectVerdict by constructing from the same registry.

**Confidence note**
This is a maintainability finding rather than a behavior claim; the current set of checks is internally consistent, but the structure is fragile for the next contract change.

**For Agents**
Look at src/verify.ts. Preserve the current verdict fields and deterministic output order. Replace CHECK_NAMES + REASONS + ad hoc checks assembly with an exact Record keyed by CompletenessChecksV1, or a finalizeChecks helper that takes the checks object once, derives ordered names from one const registry, and rejects missing entries at compile time.

File: devnet/rfc64-gate2-multi-asset-completeness/src/verify.ts [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+    raw.authored.catalogHeadDigest === raw.received.catalogHeadDigest;
+  const headCountMatchesSignedRows =
+    BigInt(raw.authored.catalogHeadTotalRows) === BigInt(signed.length);
+  const bucketCountMatchesSignedRows =
+    BigInt(raw.authored.signedBucketRowCount) === BigInt(signed.length);
+  const receivedCountMatchesSignedRows =
+    raw.received.inventoryRowCount === signed.length;
+  const signedRowsCanonicalOrder = isStrictlyOrdered(signed);
+  const activatedRowsCanonicalOrder = isStrictlyOrdered(activated);
+  const signedRowsUnique = signedDuplicates.kaIds.length === 0
```
Comment (otReviewAgent) [inline devnet/rfc64-gate2-multi-asset-completeness/src/verify.ts:90]: 🟡 Issue: Signed-row duplicate rejection is not directly covered

**What's wrong**
The verifier now distinguishes `signedRowsUnique` from `activatedRowsUnique`, but the added tests only exercise duplicate rows on the activated side. That leaves an important exact-set invariant unverified: authored signed inventories with duplicate IDs or UALs could stop being rejected without the current tests going red.

**Example**
A regression that dropped `signedDuplicates` from `signedRowsUnique` would still pass the current duplicate tests. Add a case like: duplicate or duplicate-UAL one `raw.authored.signedRows` entry, keep the artifact otherwise well formed enough to parse, then assert `checks.signedRowsUnique === false` and the duplicate ID/UAL is reported.

**Suggested direction**
Add regression tests for duplicate `kaId` and duplicate `kaUal` in `authored.signedRows`, not only in `received.activatedRows`.

**For Agents**
Look at `devnet/rfc64-gate2-multi-asset-completeness/test/completeness.test.ts` around the duplicate coverage. Add authored-side duplicate kaId and duplicate UAL cases that prove `verify()` rejects duplicates before signed-row set collapse, while preserving existing activated-side duplicate coverage.

File: devnet/rfc64-gate2-multi-asset-completeness/src/cli/verify.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    text = readBoundedUtf8(argv[2]);
+  } catch (cause) {
+    process.stderr.write(`cannot read bounded raw evidence: ${String(cause)}\n`);
+    process.exit(2);
+    return;
+  }
+  let parsed: unknown;
+  try {
+    parsed = JSON.parse(text) as unknown;
+    if (canonicalDocument(parsed as never) !== text) parsed = undefined;
```
Comment (otReviewAgent) [inline devnet/rfc64-gate2-multi-asset-completeness/src/cli/verify.ts:25]: 🟡 Issue: The command boundary for exact JCS input is untested

**What's wrong**
The README and package scripts make the generator/verifier commands part of the evidence contract, and the verifier CLI adds validation that is not exercised by the current tests. A regression in exact JCS+LF rejection, bounded reads, UTF-8 failure handling, or exit codes would not be caught.

**Example**
If line 25 were removed, `verify.ts` could accept pretty-printed JSON or a document without the required single trailing LF. The current tests would still pass because they never invoke the CLI with non-canonical input.

**Suggested direction**
Cover the CLI scripts through `child_process` tests rather than only testing the underlying pure functions.

**Confidence note**
I could not run the suite in this environment because the local Node rejects `--experimental-strip-types` and `tsc` is not installed, so this is based on static review of the added tests.

**For Agents**
Add child-process tests for `src/cli/generate.ts` and `src/cli/verify.ts`: canonical generated input exits 0 and emits canonical verdict JSON; non-canonical-but-valid JSON exits 1 with `schemaWellFormed` false; invalid count exits 2; oversized or invalid UTF-8 input exits 2.

**Comment by branarakic** [comment]
Closing — this PR's work is already on `main`, landed under different commit(s). Merging it now would **regress** `main`, because main has since evolved past the version on this branch.

**Evidence** (verified 2026-08-06 against `main` @ `0277d82c6`):
- Landed as `d5e83585a` "test(devnet): add RFC-64 Gate 2 multi-asset completeness evidence contract" and `fe3a31435` "fix(devnet): bind Gate 2 evidence to production" — both ancestors of `main`.
- **The contributed subtree is hash-identical**: `devnet/rfc64-gate2-multi-asset-completeness` is tree `ae4163a1450f202ae8bf2c7118267fbfd669b019` on both this branch and `fe3a31435`.
- Merging would delete main's workspace dependencies in `package.json` — a strict regression.

Adjudicated independently and then re-checked by a second reviewer instructed to refute the claim; it survived. Closed as part of a sweep of open PRs whose code had already landed. Please reopen if this is wrong.
